### PR TITLE
Prevent inheritance of bundle channels

### PR DESCRIPTION
**Pull request by SimonRichardson**
Bundles can have their own channels that are separate from the
application charm channels. This feature was implemented in PR#10480[1].
When moving the deployer code out from the application package into a
more testable setup we ended up causing a regression.

If a user wants to specify their own channels for each charm then they
should pin each charm correctly inside that bundle and not inherit that
channel from the bundle.

The fix is really simple, we just don't pass the origin through and
expect each charm to correctly resolve itself.

 1. https://github.com/juju/juju/pull/10480

## QA steps

Setup microk8s

```sh
sudo snap install microk8s --classic 
sudo usermod -a -G microk8s $USER
su - $USER

microk8s status --wait-ready
microk8s.enable dns storage

make microk8s-operator-update
juju bootstrap microk8s micro
```

The following would previously fail without this patch, running this will
make this work now.

```sh
juju deploy cs:kubeflow-lite
```

## Bug reference

https://bugs.launchpad.net/juju/+bug/1902945

### CODE REVIEW

File: cmd/juju/application/deployer/bundle.go [hunk 1]
```diff
@@ -215,7 +215,11 @@ type charmstoreBundle struct {
 
 // String returns a string description of the deployer.
 func (d *charmstoreBundle) String() string {
-	return fmt.Sprintf("deploy charm store bundle: %s", d.bundleURL.String())
+	str := fmt.Sprintf("deploy charm store bundle: %s", d.bundleURL.String())
+	if d.origin == (commoncharm.Origin{}) {
+		return str
+	}
+	return fmt.Sprintf("%s with origin: %s", str, d.origin.CoreChannel().String())
```
Comment (achilleasa) [inline cmd/juju/application/deployer/bundle.go:222]: Can you double-check with the other places where we log the origin and ensure we use consistent terminology across the board? IIRC we tend to use "channel: X" in this package.

Comment (SimonRichardson) [reply cmd/juju/application/deployer/bundle.go:222]: Sure can!

Comment (achilleasa) [reply cmd/juju/application/deployer/bundle.go:226]: Have a look at bundlehandler.go:338 and refresh.go:370

**Comment by SimonRichardson** [comment]
!!build!!

File: cmd/juju/application/deployer/charm.go [hunk 2]
```diff
@@ -484,3 +493,15 @@ func (c *charmStoreCharm) PrepareAndDeploy(ctx *cmd.Context, deployAPI DeployerA
 	c.origin = csOrigin
 	return c.deploy(ctx, deployAPI)
 }
+
+func isEmptyOrigin(origin commoncharm.Origin, source commoncharm.OriginSource) bool {
```
Comment (achilleasa) [inline cmd/juju/application/deployer/charm.go:497]: Why not add an `IsEmpty` to the `Origin` type? Sounds like a useful helper to have

Comment (SimonRichardson) [reply cmd/juju/application/deployer/charm.go:497]: I can do, but I'm not sure how valid it actually would be outside of this.

Comment (achilleasa) [reply cmd/juju/application/deployer/charm.go:497]: Ok, guess if it's only used here we can leave it there.

**Review by achilleasa — APPROVED** [review]
LGTM. 

We may also need to tweak the docs (and maybe the `--channel` flag description in deploy/upgrade) to make it clear how the channel is applied.

File: cmd/juju/application/deploy.go [hunk 3]
```diff
@@ -318,18 +318,24 @@ type DeployCommand struct {
...
   juju deploy cs:bionic/postgresql
-  juju deploy postgresql --series bionic
+  juju deploy cs:postgresql --series bionic
+  juju deploy cs:postgresql --channel edge
 
 All the above deployments use remote charms found in the Charm Store (denoted
-by 'cs') and therefore also make use of "charm URLs".
+by 'cs' prefix) and therefore also make use of "charm URLs".
+
+Specifying a channel source will use that as the source to look for the charm or
```
Comment (achilleasa) [inline cmd/juju/application/deploy.go:335]: Maybe reworded as:

```
If a channel is specified, it will be used as the source for looking up the charm or bundle from the Charm Store. When used in a bundle deployment context, the specified channel is only used for retrieving the bundle and is ignored when looking up the charms referenced by the bundle. However, each charm within a bundle is allowed to explicitly specify the channel used to look it up.
```

Or something along those lines?

Comment (SimonRichardson) [reply cmd/juju/application/deploy.go]: Yes!

**Comment by SimonRichardson** [comment]
$$merge$$

**Comment by achilleasa** [comment]
QA works as expected
